include/cache.h

```diff
@@ -13,57 +13,178 @@
 extern "C" {
 #endif
 
-void arch_dcache_flush(void *addr, size_t size);
-void arch_dcache_invd(void *addr, size_t size);
+/**
+ * Common operations for the caches
+ */
+#define K_CACHE_INVD		BIT(1)
+#define K_CACHE_FLUSH		BIT(2)
+#define K_CACHE_FLUSH_INVD	(K_CACHE_FLUSH | K_CACHE_INVD)
+#define K_CACHE_CLEAN		BIT(3)
```

Comment:
At least on AArch64 you can have the three operations: flush (`dc civac`), clean (`dc cvac`) and invalidate (`dc ivac`) so I'm trying to consider all the possible cases here.

Reply:
Hi @carlocaione , could you please confirm what is a "clean" operation meaning? Looking through ARM docs, it seems "clean" is an ARM term for "flushing" (which is used by ARC, for example), so clean==flush.
Regarding instructions that you listed, ARMv8 ARM defines them as:
DC CVAC - Clean by virtual Address
DC CIVAC - Clean and invalidate by VA
DC IVAC - Invalidate by VA

Reply:
Hi @iMitsyanko. AFAIK in the AArch64 world usually the `flush` operation is `clean + invalidate`, see for example https://github.com/ARM-software/arm-trusted-firmware/blob/master/lib/aarch64/cache_helpers.S#L43-L44 or https://github.com/u-boot/u-boot/blob/master/arch/arm/cpu/armv8/cache.S#L132-L142 while the `clean` operation without `invalidate` is a different operation.

So, for AArch64 we have:
- `clean`
- `invalidate`
- `flush == clean + invalidate`

I decided to keep flush as standalone operation because I don't know how these are mapped on other architectures so I tried to be as general as possible with `clean`, `invalidate` and `flush` as three distinct operations. Each arch can map the three ops as they please.

Reply:
This definitely makes sense, though it seems in that case K_CACHE_FLUSH_INVD and K_CACHE_CLEAN_INVD definitions should be removed from this PR?

Reply:
No objection to remove those. I can leave only `K_CACHE_INVD` `K_CACHE_FLUSH` and `K_CACHE_CLEAN` and each arch can `OR` them if they want to.

Reply:
One thing to note : Most of the latest armv8 cores do not really differentiate the clean and invalidate CMOs. ie. DC IVAC (invalidate) is promoted to DC CIVAC (clean and invalidate).

From OS perspective if we have architecture agnostic range based flush for I$ D$ that should be sufficient. Do you have a use case in mind for the granular `clean` operation. How it will be used ?
Rest of the APIs looks good to me.

Reply:
I don't have in mind any use case but we cannot exclude that some architecture can actually fully differentiate between `flush` `clean` and `invalidate`, so I'm just trying to be a bit fool / future proof here leaving all the options on the table.

Reply:
@carlocaine IMHO we should not use ARM nomenclature (`flush`/`invalidate`/`clean`) which might be misleading on non-ARM machines, but instead name the operations basing on what they do: `write_back`, `invalidate` and `write_back_invalidate`.

Reply:
Interesting. Is that really ARM nomenclature? At least ARC, X86 and NIOS2 seem to use `flush` with a similar meaning. I'm not strongly opinionated on the naming so I'll change it for the sake of moving this forward (but admittedly I like `flush` more than `write_back`)

Reply:
Note that the proposed names can be shorted to wb/inv/wb_inv to make lazy programmers happy :)

Reply:
updated ;)